homeassistant/helpers/label_registry.py

```diff
@@ -0,0 +1,244 @@
+"""Provide a way to label and categorize anything."""
+from __future__ import annotations
+
+from collections.abc import Iterable, MutableMapping
+from typing import cast
+
+import attr
+
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.util import slugify
+
+from . import device_registry as dr, entity_registry as er
+from .typing import UNDEFINED, UndefinedType
+
+DATA_REGISTRY = "label_registry"
+EVENT_LABEL_REGISTRY_UPDATED = "label_registry_updated"
+STORAGE_KEY = "core.label_registry"
+STORAGE_VERSION_MAJOR = 1
+SAVE_DELAY = 10
+
+
+@attr.s(slots=True, frozen=True)
+class LabelEntry:
+    """Label Registry Entry."""
+
+    label_id: str = attr.ib()
+    name: str = attr.ib()
+    normalized_name: str = attr.ib()
+    description: str | None = attr.ib(default=None)
+    color: str | None = attr.ib(default=None)
+    icon: str | None = attr.ib(default=None)
+
+
+class LabelRegistry:
+    """Class to hold a registry of labels."""
+
+    def __init__(self, hass: HomeAssistant) -> None:
+        """Initialize the label registry."""
+        self.hass = hass
+        self.labels: MutableMapping[str, LabelEntry] = {}
+        self._store = hass.helpers.storage.Store(
+            STORAGE_VERSION_MAJOR,
+            STORAGE_KEY,
+            atomic_writes=True,
+        )
+        self._normalized_name_label_idx: dict[str, str] = {}
+        self.children: dict[str, set[str]] = {}
+
+    @callback
+    def async_get_label(self, label_id: str) -> LabelEntry | None:
+        """Get label by id."""
+        return self.labels.get(label_id)
+
+    @callback
+    def async_get_label_by_name(self, name: str) -> LabelEntry | None:
+        """Get label by name."""
+        normalized_name = normalize_label_name(name)
+        if normalized_name not in self._normalized_name_label_idx:
+            return None
+        return self.labels[self._normalized_name_label_idx[normalized_name]]
+
+    @callback
+    def async_list_labels(self) -> Iterable[LabelEntry]:
+        """Get all labels."""
+        return self.labels.values()
+
+    @callback
+    def async_get_or_create(self, name: str) -> LabelEntry:
+        """Get or create an label."""
+        if label := self.async_get_label_by_name(name):
+            return label
+        return self.async_create(name)
+
+    @callback
+    def _generate_id(self, name: str) -> str:
+        """Initialize ID."""
+        suggestion = suggestion_base = slugify(name)
+        tries = 1
+        while suggestion in self.labels:
+            tries += 1
+            suggestion = f"{suggestion_base}_{tries}"
+        return suggestion
```

Comment:
I dunno, I kept it the same as area ID, as this is part of a discussion that felt endless/has loose ends on areas too.

When I set a different solution here in labels, it would be inconsistent with areas. Or would imply I would have to change labels too (but that discussion never finished either).

TL;DR: I didn't want to get myself into that discussion and kept it consistent.